### ISSUE TITLE
Creating a free-standing extension ends in a segfault

### ISSUE THREAD

**Issue by lukaslueg** [rusqlite/rusqlite#524]
I figured it should be possible to create a free-standing extension (a `cdylib`) for sqlite using the rusqlite-API. The basic plan is to

* have a small C-shim, which does the extension-initialization via the macros provided by sqlite. This can't be done in Rust, as those macros are not callable via a binding.
* have any sqlite-client load the extension, which calls the shim, which calls some Rust-code
* ???
* profit

I've implemented a small extension which does the above to provide a new aggregate-function. The basic loading mechanism seems to work fine, yet sqlite crashes with a segfault somewhere in [here](https://github.com/lukaslueg/rusqlite_ext_test/blob/6b46f025169032f714c942de692b7c5b3d71ceb3/src/lib.rs#L24) when calling `sqlite3_create_function_v2`. To reproduce, build the lib and use `.load target/debug/libsqlite3_hyperminhash.so` in  the sqlite-cli.

I was unable to debug this any further as of yet. Am I missing something obvious here?

**Comment by thomcc** [comment]
Unfortunately, the SQLite extension API also re-`#defines` a bunch of normal SQLite methods to call through a provided API pointer, which is done in case the SQLite version you compile against and are loaded into do not match (and so that you don't actually have to link against SQLite).

I investigated this before and concluded that it wasn't really practical to use Rusqlite with a SQLite extension, but I also didn't try very hard.

**Comment by lukaslueg** [comment]
I've checked out what the INIT-macros actually do and stopped investigating this any further... Fine with me to close

**Comment by jrandall** [comment]
I can confirm that #532 should address this issue. 

> have a small C-shim, which does the extension-initialization via the macros provided by sqlite. This can't be done in Rust, as those macros are not callable via a binding.
@lukaslueg with #532 I think it should not be necessary to use a C-shim at all. You should be able to use it to compile a rust cdylib that can be directly loaded as a sqlite extension. 

I'll try to explain here how it works and how to use it in case you want to give it a try. 

The way this works is that there are multiple sets of rust bindings generated for sqlite, one for the normal case and others for the case when the rusqlite code is being used as an extension. This basically accomplishes the same thing that including `sqlite3.h` vs `sqlite3ext.h` in your C code would accomplish. 

The normal rusqlite/libsqlite3 bindings generate an empty/opaque `sqlite3_api_routines` struct:
https://github.com/jgallagher/rusqlite/blob/b21aed20f037dbc3b2cff06bf4262ba1ff488099/libsqlite3-sys/bindgen-bindings/bindgen_3.26.0.rs#L834

The extension version instead generates a fully populated `sqlite3_api_routines` struct:
https://github.com/jgallagher/rusqlite/blob/b21aed20f037dbc3b2cff06bf4262ba1ff488099/libsqlite3-sys/bindgen-bindings/bindgen_3.26.0-ext.rs#L3574-L4938

The need to call the various `SQLITE_EXTENSION_INIT` preprocessor macros is replaced by adjusting the generated bindings that are built during the build phase (or included in the prebuilt extension bindings). The `SQLITE_EXTENSION_INIT1` macro defines and initialise to null a `sqlite3_api` global to point to an instance of the `sqlite3_api_routines` struct. Elsewhere in `sqlite3ext.h`, preprocessor definitions are used to override all of the actual sqlite3 api functions with macros that call function pointers to the routines accessed via the `sqlite3_api` global instead. 

The extension bindings in #532 define (and initialise to null) a static `sqlite3_api` variable (this replaces the function of the `SQLITE_EXTENSION_INIT1` macro):
https://github.com/jgallagher/rusqlite/blob/b21aed20f037dbc3b2cff06bf4262ba1ff488099/libsqlite3-sys/bindgen-bindings/bindgen_3.26.0-ext.rs#L7755-L7759

The non-extension bindings declare the sqlite3 api functions directly as `extern "C"` functions (e.g. `sqlite3_open_v2`: https://github.com/jgallagher/rusqlite/blob/b21aed20f037dbc3b2cff06bf4262ba1ff488099/libsqlite3-sys/bindgen-bindings/bindgen_3.26.0.rs#L1525-L1532), while the extension version suppresses the generation of direct function bindings for all of those functions and instead generates a rust wrapper function that dispatches to the function pointer stored in the `sqlite3_api` static variable (again e.g. `sqlite3_open_v2`: https://github.com/jgallagher/rusqlite/blob/b21aed20f037dbc3b2cff06bf4262ba1ff488099/libsqlite3-sys/bindgen-bindings/bindgen_3.26.0-ext.rs#L9687-L9701). 

While this may seem more complicated than the C version, all that is really happening is that, while in the C code the preprocessor effectively rewrites the interface code to replace direct api calls with those dispatched through the global `sqlite3_api` instance, in the rust code the equivalent is to simply generate different versions of the binding code and choose which one to use during build time (in the build script). 

The two remaining init macros are `SQLITE_EXTENSION_INIT2(v)` and `SQLITE_EXTENSION_INIT3` (although I would note that the latter is underdocumented). In C, these offer two different ways to set the `sqlite3_api` global to a real value that points to a valid `sqlite3_api_routines` struct that has been populated by sqlite rather than the null value it is initialized with. 

The `SQLITE_EXTENSION_INIT2(v)` is intended to be called (once) from within an extension's init function (i.e. `sqlite3_extension_init` or a version named after your library such as `sqlite3_mylib_init`). The signature of the extension init function includes being passed a valid sqlite3_api_routines struct in `pApi`. In C, we call `SQLITE_EXTENSION_INIT2(pApi)`. For our rust extension to work, our rust code must do the same thing. More on that below.

The `SQLITE_EXTENSION_INIT3` macro is an alternative to `SQLITE_EXTENSION_INIT2` (i.e. you use one or the other but not both). It is used when you have code that needs to talk to sqlite from within an extension (i.e. linked into an extension), but which does not itself include the extension entrypoint (i.e. because it is compiled into a separate object or library file). 

In `sqlite3ext.h` this is defined as:
```C
# define SQLITE_EXTENSION_INIT3     \
    extern const sqlite3_api_routines *sqlite3_api;
```
The idea is that the code that calls `SQLITE_EXTENSION_INIT3` is embedded within (i.e. linked into) a sqlite extension, and so the `sqlite3_api` global is an exterrn that is expected to be provided by the extension itself (which will have called `SQLITE_EXTENSION_INIT2` when it is initialized). 

Both the non-embedded and the embedded scenarios are supported by #532. The latter use case would be when you would have used `SQLITE_EXTENSION_INIT3` in C, and it is handled by the `-ext-embed` variant of the bindings which includes a declaration of `sqlite3_api` as an extern: https://github.com/jgallagher/rusqlite/blob/b21aed20f037dbc3b2cff06bf4262ba1ff488099/libsqlite3-sys/bindgen-bindings/bindgen_3.26.0-ext-embed.rs#L7758-L7761

This would be useful if you have an existing C/C++ sqlite extension but you want to write a rust library that adds functionality to it. For example, I am working on a project in which a single sqlite extension implements a number of different virtual tables. In our project, the main extension is implemented in C, but embeds virtual tables that are implemented in C, C++, and Rust and linked into the main extension. 

If you compile your rust library as a `cdylib` using rusqlite with the `loadable_extension_embedded` feature, the embedded variant of the bindings will be used and you will be able to call (most) sqlite api functions as normal from within the rust code once it is linked into another library that provides and initializes an extern `sqlite3_api` symbol. 

The more straightforward non-embedded case is when you would have used the `SQLITE3_EXTENSION_INIT2(v)` macro in C. In other words, when you want your rust library to itself be a standalone sqlite extension that can be called from any sqlite environment (including the CLI). In that case, you should use rusqlite with the `loadable_extension` feature implemented in #532 . For it to work, you must also provide a C function that implements the extension entrypoint (init function), and from within that you must set the `sqlite3_api` global to the `p_api` value it provides. 

For example:
```Rust
#[no_mangle]
pub extern "C" fn sqlite3_extension_init(
    db: *mut ffi::sqlite3,
    pz_err_msg: &mut &mut std::os::raw::c_char,
    p_api: *mut ffi::sqlite3_api_routines,
) -> c_int {
    // SQLITE_EXTENSION_INIT2 equivalent
    unsafe {
        ffi::sqlite3_api = p_api;
    }
    // Do some things to init your vtable here
    return ffi::SQLITE_OK;
}
```

**Comment by lukaslueg** [comment]
Thanks for the heads-up. I put #532 in my watchlist and will port [`sqlite3_hyperminhash`](https://github.com/lukaslueg/sqlite3_hyperminhash) to purely use rusqlite once it got merged.
